### DISCUSSION TITLE
Autonomous Vehicle Localization Using Onboard Sensors and HD Geolocated Maps (Project 20)

### DISCUSSION THREAD

**Discussion by robertogl** [mathworks/MATLAB-Simulink-Challenge-Project-Hub#3] [Collaborate]
Contribute to the discussion by asking and/or answering questions, commenting, or sharing your ideas for solutions to [project #20](https://github.com/mathworks/MathWorks-Excellence-in-Innovation/tree/main/projects/Autonomous%20Vehicle%20Localization%20Using%20Onboard%20Sensors%20and%20HD%20Geolocated%20Maps)

---
*:robot: An AI bot may reply to your comments with suggestions and guidance. It will not solve the project for you. Your comments may be processed by a third-party LLM. To direct a question to a specific person, tag them with @username.*

**Comment by NipunAgarwal16** [comment]
Hi, I am Nipun Agarwal, a student at BITS PILANI and interested in working and collaborating on this project.

**Comment by emcca029** [comment]
Hi, I am a student at FIU and am interested in working and collaborating on this project.

**Reply by Krishram97** [reply]
have got /solve the first Automation of vehicle

**Comment by Krishram97** [comment]
Hi, I am Masters from JNTUH and interested in working and collaborating on this project

**Comment by AkuG18O2** [comment]
Hi, I am an undergrad at IIT Delhi and interested in working and collaborating on this project. I am planning on starting with a Stereo camera and INS setup, in an Unreal simulation environment. I'll consider adding a lidar source as well once the initial flow is established. Any tips to start off?

**Reply by robertogl** [reply]
Hello @AkuG18O2,

Great to hear you will be working on this project!
We have a large number of [examples ](https://www.mathworks.com/help/driving/examples.html) shipped with the Automated Driving Toolbox that tackle different challenges of autonomous vehicles. You can start by getting familiar with the ones that are closer to the approach you are adopting and even use one of them as your starting point. Please let us know if you need anything else.

Have fun!

**Reply by AkuG18O2** [reply]
Hi. So I was currently working on lidar SLAM and am having a problem. I was trying to edit the point cloud output from 3d lidar block using a Matlab function block in simulink, but after processing I am unable to return the processed point cloud from the function. It gives an error saying the point cloud is a class object and not a matrix or scalar, hence the problem. In most examples I have looked at, the lidar data has been recorded for the entire scenario beforehand from the simulink model and the slam algorithm is run afterwards separately using pure MATLAB scripts.

Any advice on how I can go about simultaneously getting lidar data and processing it into a map?

The most promising approach I have found yet (from this example : https://in.mathworks.com/help/fusion/ug/tracking-vehicles-using-lidar-in-simulink.html) seems to be to use MATLAB System blocks. But I don't have much experience using systems (only used functions previously) so it may take some time getting used to.

Is this my best bet or is there a better method to process the lidar data into a point cloud in the same Simulink model?

**Reply by AkuG18O2** [reply]
Okay so I have figured it out to some extent using persistent variables and extrinsic functions. Let's see...

**Reply by robertogl** [reply]
I was actually looking at it now. I am glad you solved it, but let us know if there is more.

**Reply by robertogl** [reply]
Let me add some suggestions anyway. After processing, rather than returning the point cloud as an object, you can extract just the XYZ points and have that array as an output. If needed, you can reorganize the point cloud later on. Beside the XYZ points you might need to extract and output some other properties from the same object.

**Reply by AkuG18O2** [reply]
Yeah that's what I ended up doing as well :) Thanks for the tip.

**Comment by Krishram97** [comment]
You have to start with a camera then with a stereo camera then you can
collaborate with INS
   K R K

**Reply by sskulka** [reply]
@Krishram97 did you implement orb slam 3 for the INS fusion?

**Comment by Abetelgeusian** [comment]
@robertogl We started the work on this project. However, there is an issue with HERE HDLM. Even if you create an account, and somehow generate the ID and secret key, it doesn't work. I contacted the HERE support, and they said this feature is not available to the public yet.
He also said there might be an issue with licensing between MathWorks and HERE.
Could you please advise on how to proceed with using HERE HDLM?

**Reply by robertogl** [reply]
I guess that happens when you use the [hereHDLMReader](https://www.mathworks.com/help/driving/ref/herehdlmreader.html) function, right? Let me investigate.
What MATLAB version are you using?

**Reply by Abetelgeusian** [reply]
Yes. I am using MATLAB R2021a.

**Reply by robertogl** [reply]
Were you at least able to validate your credentials? If so, what error are you getting?

**Reply by Abetelgeusian** [reply]
Apologies for the delayed response. 
No, I cannot even log in. (When I generated the access key id and access secret key, I stored them in my system.) Whenever I use these credentials, it says unable to validate credentials.  Is there a way to know that the credentials I have are the right ones?

**Reply by robertogl** [reply]
Once you create the credentials you should be able to download the file 'credentials.properties'. You have to move this file inside the folder named '.here' that you were supposed to create inside your HOMEPATH (just write the command 'cd %HOMEPATH%' in the command prompt). Then copy the credentials from the 'credentials.properties' file (open it with notepad) and use those two keys in the MATLAB dialog box. That should at least work for enabling the credentials, did you follow this process already?

**Reply by sskulka** [reply]
@robertogl Hello, I am project mate of @Abetelgeusian working on the project. We were able to since login suing the credentials but we are getting an error while creating the hdlm configuration object. I tried with 2 seperate zones but both are throwing up the same issue.
![hdlm_config_forbidden_error](https://user-images.githubusercontent.com/90679857/165218806-1d4b97d1-e30f-43e6-a381-7367834d7d1e.PNG)
.

**Reply by robertogl** [reply]
Hi @sskulka, 

Now you have to associate the app/credential to the data you want to access. Under “Data” in the Marketplace portal, you will see all of the available catalogs. If your account has access to the HD Live Map data, you will see it listed here. There should be two catalogs with the HERE HD Live Map Native (or Protobuf) name: one for North America and one for Western Europe. If you see these, you can click on each and under “Sharing” you can add which apps you want to give access to that catalog. If you do not see these under “Data” then you would need to request access to these data from HERE, either via their Marketplace interface or contacting their customer rep directly.

**Reply by sskulka** [reply]
Hello @robertogl ,
Thanks for the directions. I checked on the marketplace but I wasn't able to see the catalogs under Data. I raised a ticket with their support and here's what they had to say about this.

Do you know if we can create a commercial account with Matlab or access an existing account since we're doing the project. Also, can we please schedule a short call with you since the project pipeline is becoming a bit ambiguous for us since we're not sure if we can use the here hdlm. Could we discuss the options to progress for creating the scenario along with GPS data?

**Reply by robertogl** [reply]
Hi @sskulka,
Thank you for letting me know. Let me discuss this internally and see if there is anything we can do about the commercial account. In the meantime let me contact you directly by email.

**Reply by AkuG18O2** [reply]
I suppose the HERE map integration is only optional for the SLAM algorithm right? At least the scenario can be designed using just the 3D environment in Unreal engine. Please correct me if I am mistaken.

**Reply by robertogl** [reply]
You are absolutely right! And thanks for bringing that up.

**Reply by sskulka** [reply]
@AkuG18O2 you're correct...we were initially focusing on the GPS-based localisation through sign matching using the object detection pipeline. There's an interesting matlab reference for that. But we ended up doing slam anyways.

**Comment by sskulka** [comment]
![Screenshot_20220427-140957](https://user-images.githubusercontent.com/90679857/165604860-fbcfe5cd-ef97-4229-bbb4-ed7692d1ab1c.png)

**Comment by emcca029** [comment]
Hello,

I was wondering if you guys have room for 1 more person on your project?

**Reply by robertogl** [reply]
Hi @emcca029,

Are you asking someone in particular? If you are asking me, we have unlimited space :)!

**Comment by Krishram97** [comment]
I have started with new build with Automated driving toolbox in MATLAB so I
need some time to make it. Please help how to connect unity with MATLAB
   K R K



---------- Forwarded message ---------
From: robertogl ***@***.***>
Date: Thu, Mar 17, 2022 at 7:00 PM
Subject: Re: [mathworks/MathWorks-Excellence-in-Innovation] Autonomous
Vehicle Localization Using Onboard Sensors and HD Geolocated Maps (Project
20) (Discussion #3)
To: mathworks/MathWorks-Excellence-in-Innovation <
***@***.***>
Cc: Krishram_97 ***@***.***>, Comment ***@***.***
>


Hello @AkuG18O2 <https://github.com/AkuG18O2>,

Great to hear you will be working on this project!
We have a large number of examples
<https://www.mathworks.com/help/driving/examples.html> shipped with the
Automated Driving Toolbox that tackle different challenges of autonomous
vehicles. You can start by getting familiar with the ones that are closer
to the approach you are adopting and even use one of them as your starting
point. Please let us know if you need anything else.

Have fun!

—
Reply to this email directly, view it on GitHub
<https://github.com/mathworks/MathWorks-Excellence-in-Innovation/discussions/3#discussioncomment-2380808>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/ALKCRBI2JNSV2CZVPQH7JOLVAMXYZANCNFSM4ZNHA7OQ>
.
Triage notifications on the go with GitHub Mobile for iOS
<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
or Android
<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

You are receiving this because you commented.Message ID:
<mathworks/MathWorks-Excellence-in-Innovation/repo-discussions/3/comments/2380808
@github.com>

**Reply by thewitek** [reply]
Unity is not supported out of the box.  You can use a built-in Unreal support.

Witek

**Comment by Priyadharshan2003** [comment]
Hi, I am a Priyadharshan C , student at Sona  College of Technology and interested in working and collaborating on this project

**Comment by Hazem-M-Abdelaziz** [comment]
Hey there, I face a problem getting the trial of Automated Driving Toolbox for student account after requesting it many times.
Is there anything else I can do ?

**Reply by robertogl** [reply]
Hi @Hazem-M-Abdelaziz,
Sorry for the late response, I sent you an email regarding this. Let's hope we can fix this issue.
Cheers!
